frontend/catalyst/compiler.py

```diff
@@ -248,13 +248,21 @@ def run_writing_command(command: List[str], compile_options: Optional[CompileOpt
         "convert-index-to-llvm",
         "convert-catalyst-to-llvm",
         "convert-quantum-to-llvm",
+        # There should be no identical code folding
+        # (`mergeIdenticalBlocks` in the MLIR source code)
+        # between convert-async-to-llvm and
+        # add-exception-handling.
+        # So, if there's a pass from the beginning
+        # of this list to here that does folding
+        # add-exception-handling will fail to add async.drop_ref
+        # correctly. See https://github.com/PennyLaneAI/catalyst/pull/995
```

Comment:
Is this is a limitation of of how we specify the async pipeline? As in would we otherwise be able to run the passes closer together which would mitigate this issue?

Reply:
I think there might be an order for which `--convert-async-to-llvm` may be placed later in the pipeline, but there will likely need to be some passes in between `--convert-async-to-llvm` and `--add-exception-handling`. I am not a big fan of optionally adding passes (it makes the pipeline more difficult to reason about) but I understand its advantages and disadvantages.

Going back to how we specify the async pipeline, right now we place the async conversion at the top just as a convenience, but there's no reason why we can't specify it at a particular offset. But it would be also less readable to do so.

Reply:
Definitely using an offset would be very awkward. It would be much better if we could specify optional passes directly inline, something like:
```py
        "convert-catalyst-to-llvm",
        "convert-quantum-to-llvm",
        "convert-async-to-llvm" if config.async else None,  # or `else ""`
        "add-exception-handling",
        "emit-catalyst-py-interface",
```

Reply:
It should be possible if we just make sure to remove None's in the pipeline.

Reply:
Not sure if you want this for this PR or another PR.

Reply:
Up to you :)

Reply:
Another PR. I will add it to #997